src/main/java/com/google/firebase/auth/ListUsersPage.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.gax.paging.Page;
+import com.google.common.collect.ImmutableList;
+import com.google.firebase.auth.internal.DownloadAccountResponse;
+import com.google.firebase.internal.NonNull;
+import com.google.firebase.internal.Nullable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * Represents a page of {@link ExportedUserRecord} instances. Provides methods for iterating
+ * over the users in the current page, and calling up subsequent pages of users. Instances of
+ * this class are thread safe and immutable.
+ */
+public class ListUsersPage implements Page<ExportedUserRecord> {
+
+  static final String END_OF_LIST = "";
+
+  private final UserSource source;
+  private final int maxResults;
+
+  private final ListUsersResult currentBatch;
+
+  private ListUsersPage(@NonNull UserSource source, int maxResults, @Nullable String pageToken) {
+    this.source = source;
+    this.maxResults = maxResults;
+    this.currentBatch = source.fetch(maxResults, pageToken);
```

Comment:
This is performing a synchronous network operation inside the constructor. Can you move this code to the accessors?

Reply:
This class is much easier to reason about when it's immutable. Therefore I'd rather fetch the data upon construction, and assign to a `final` attribute. Also lazy loading is not a requirement for this class. `FirebaseAuth.listUsers()` and `Page.getNextPage()` are expected to make RPC calls, and return fully populated pages as the result.

We can move the RPC call out of the constructor, but since we create pages in 2 places the constructor seems like a natural place to place that bit of code. Alternatively we can try a private `ensureCurrentBatch()` method to load the data on demand, but we are not really saving anything since all operations on `Page` will have to call it.

Reply:
This is generally frowned upon as it makes testing hard (SeeL http://misko.hevery.com/code-reviewers-guide/flaw-constructor-does-real-work/).

One easy way to mitigate would be to create a factory method and pass in `currentBatch`.